libtenzir/builtins/operators/sample.cpp

```diff
@@ -0,0 +1,214 @@
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/arrow_utils.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/series_builder.hpp>
+#include <tenzir/tql2/plugin.hpp>
+
+#include <arrow/compute/api.h>
+
+namespace tenzir::plugins::sample_ {
+
+namespace {
+
+enum class mode { none, log, log2, log10, sqrt };
+
+class sample_operator final : public crtp_operator<sample_operator> {
+public:
+  sample_operator() = default;
+
+  sample_operator(mode m, duration delta) : mode_{m}, delta_{std::move(delta)} {
+  }
+
+  auto operator()(generator<table_slice> input,
+                  operator_control_plane&) const -> generator<table_slice> {
+    // XXX: Decide if the window should keep rolling even if no events are
+    // received.
+    auto last = std::chrono::steady_clock::now();
+    auto count = uint64_t{};
+    // Logic copied from the slice_operator::positive_stride()
+    // TODO: Consider using the slice operator directly or extracting this
+    // funcitonality.
+    auto offset = int64_t{};
+    auto stride = int64_t{1};
+    const auto calc = [&]() {
+      if (count == 1) {
+        return 1.0;
+      }
+      switch (mode_) {
+        case mode::log:
+          return std::log(count);
+        case mode::log2:
+          return std::log2(count);
+        case mode::log10:
+          return std::log10(count);
+        case mode::sqrt:
+          return std::sqrt(count);
+        case mode::none:
+          TENZIR_UNREACHABLE();
+      }
+      TENZIR_UNREACHABLE();
+    };
+    constexpr auto make_stride_index
+      = [](const int64_t offset, const int64_t rows, const int64_t stride) {
+          TENZIR_ASSERT(stride > 0);
+          auto b = int64_type::make_arrow_builder(arrow::default_memory_pool());
+          check(b->Reserve((rows + 1) / stride));
+          for (auto i = offset % stride; i < rows; i += stride) {
+            check(b->Append(i));
+          }
+          return finish(*b);
+        };
+    for (const auto& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      // XXX: >=?
+      if (auto now = std::chrono::steady_clock::now(); now - last > delta_) {
+        stride = std::ceil(calc());
+        last = now;
+        offset = 0;
+        count = 0;
+      }
+      count += slice.rows();
+      auto batch = to_record_batch(slice);
+      const auto rows = batch->num_rows();
+      auto stride_index = make_stride_index(offset, rows, stride);
+      offset += rows;
+      auto take_result = arrow::compute::Take(batch, stride_index);
+      if (not take_result.ok()) {
+        // XXX: Should this be an error?
+        diagnostic::error("{}", take_result.status().ToString())
+          .note("failed to apply stride")
+          .throw_();
+      }
+      const auto datum = take_result.MoveValueUnsafe();
```

Comment:
```suggestion
      const auto datum = check(arrow::compute::Take(batch, stride_index));
```

Don't think this can really fail outside of logic errors (for which an assertion is fine) and allocation failures (which we cannot handle anyways).